tap_intercom/stream.py

```diff
@@ -16,34 +17,49 @@ def __init__(self, config: Dict):
     def do_sync(self, tap_stream_id: str, state: Optional[datetime] = None):
 
         prev_bookmark = None
+        start_date, end_date = self.__get_start_end(
+            state=state, bookmark_key=BOOKMARK_KEY, tap_stream_id=tap_stream_id
+        )
         with singer.metrics.record_counter(tap_stream_id) as counter:
 
             try:
                 data = self.intercom.get_records(tap_stream_id)
                 for record, replication_value in data:
 
-                        if replication_value and (
-                            start_date >= replication_value or end_date <= replication_value
-                        ):
-                            continue
-
-
-
-                        new_bookmark = replication_value
-                        if not prev_bookmark:
-                            prev_bookmark = new_bookmark
-
-                        if prev_bookmark < new_bookmark:
-                            state = self.__advance_bookmark(state, prev_bookmark)
-                            prev_bookmark = new_bookmark
-
-                    return self.__advance_bookmark(state, prev_bookmark)
-
-                except Exception:
-                    self.__advance_bookmark(state, prev_bookmark)
-                    raise
-
-    def __get_start_end(self, state: dict):
+                    if replication_value and (
+                        start_date >= replication_value or end_date <= replication_value
+                    ):
+                        continue
+
+
+                    singer.write_record(tap_stream_id, record)
+                    counter.increment(1)
+                    if not replication_value:
+                        continue
+
+                    new_bookmark = replication_value
+                    if not prev_bookmark:
+                        prev_bookmark = new_bookmark
+
+                    if prev_bookmark >= new_bookmark:
+                        continue
+                    self.__advance_bookmark(
+                        state=state,
+                        bookmark=prev_bookmark,
+                        bookmark_key=BOOKMARK_KEY,
+                        tap_stream_id=tap_stream_id,
+                    )
+                    prev_bookmark = new_bookmark
+
+            finally:
+                self.__advance_bookmark(
+                    state=state,
+                    bookmark=prev_bookmark,
+                    bookmark_key=BOOKMARK_KEY,
```

Comment:
If `bookmark_key` is a global value `BOOKMARK_KEY` then we don't need to pass it as argument, we can just use it in `__advance_bookmark` function :)